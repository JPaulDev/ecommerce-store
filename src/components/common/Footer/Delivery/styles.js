import styled, { css } from 'styled-components';

export const Container = styled.div`
  @media screen and (min-width: 650px) {
    margin-bottom: 30px;
    margin-right: 30px;
  }

  @media screen and (min-width: 860px) {
    margin-right: 50px;
  }

  @media screen and (min-width: 1050px) {
    margin: 0;
  }
`;

export const Text = styled.p`
  color: var(--color-grey-4);
  font-size: var(--font-size-13);
  margin: 8px 0 8px 20px;

  @media screen and (min-width: 650px) {
    margin: 8px 0;
  }
`;

export const ButtonContainer = styled.div`
  display: flex;
`;

export const TabButton = styled.button`
  background-color: ${({ active }) =>
    active ? 'var(--color-grey-11)' : 'var(--color-grey-12)'};
  border-bottom: ${({ active }) =>
    active
      ? '3px solid var(--color-orange-1)'
      : '3px solid var(--color-grey-12)'};
  cursor: pointer;
  flex: 1;
  font-weight: var(--font-weight-semibold);
  padding: 8px;
  transition: all 200ms ease-in-out;
`;

export const Form = styled.form`
  align-items: center;
  background-color: var(--color-grey-11);
  border-bottom: 1px solid var(--color-grey-9);
  display: flex;
  padding: 20px 15px;

  @media screen and (max-width: 649px) {
    justify-content: center;
  }
`;

export const inputStyles = css`
  border: 1px solid grey;
  border-radius: 5px;
  margin-right: 8px;
  max-width: 175px;
  padding: 8px 0 8px 10px;
  width: 40%;
`;

export const SubmitButton = styled.button`
  align-items: center;
  background: var(--color-grey-3);
  border-radius: 50%;
  cursor: pointer;
  display: flex;
  height: 32px;
  justify-content: center;
  width: 32px;

  &:active {
    transform: scale(0.9);
  }
`;
