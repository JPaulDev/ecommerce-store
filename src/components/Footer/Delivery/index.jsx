import { useState, useEffect, useId } from 'react';
import { motion, AnimatePresence } from 'framer-motion';
import useMediaQuery from '../../../hooks/useMediaQuery';
import MaterialIcon from '../../MaterialIcon';
import { Heading, DropdownButton } from '../styles';
import * as Styled from './styles';

const forms = [
  {
    action: 'https://pc-connect.co.uk/tracking/trackorder',
    placeholder: 'Postcode',
    name: 'postcode',
  },
  {
    action: 'https://pc-connect.co.uk/tracking/trackreturn',
    placeholder: 'RMA No.',
    name: 'rmaNumber',
  },
  {
    action: 'https://pc-connect.co.uk/tracking/trackbuild',
    placeholder: 'Build No.',
    name: 'buildNumber',
  },
];

export default function Delivery() {
  const [activeTab, setActiveTab] = useState(0);
  const [isOpen, setIsOpen] = useState(window.innerWidth >= 650);
  const isMatch = useMediaQuery('(min-width: 650px)');
  const id = useId();
  const heading = 'Delivery / Tracking Information';

  const handleSetActiveTab = (index) => setActiveTab(index);
  const handleOpen = () => setIsOpen(!isOpen);

  // Prevents the menu being stuck closed when resizing from a smaller window width
  useEffect(() => {
    isMatch ? setIsOpen(true) : setIsOpen(false);
  }, [isMatch]);

  return (
    <Styled.Container>
      {isMatch ? (
        <Heading>{heading}</Heading>
      ) : (
        <DropdownButton
          type="button"
          isOpen={isOpen}
          aria-expanded={isOpen}
          aria-controls="delivery"
          onClick={handleOpen}
        >
          {heading}
          <MaterialIcon iconSize="30px">chevron_right</MaterialIcon>
        </DropdownButton>
      )}
      <AnimatePresence initial={false}>
        {isOpen && (
          <motion.div
            id="delivery"
            initial={{ height: 0 }}
            animate={{ height: 'auto' }}
            exit={{ height: 0 }}
            transition={{ ease: 'easeInOut', duration: 0.5 }}
          >
            <Styled.Text>Order now to receive by tomorrow.</Styled.Text>
            <Styled.ButtonContainer>
              {['Order', 'Return', 'System'].map((text, index) => (
                <Styled.TabButton
                  key={text}
                  active={index === activeTab}
                  onClick={() => handleSetActiveTab(index)}
                  type="button"
                >
                  {text}
                </Styled.TabButton>
              ))}
            </Styled.ButtonContainer>
            {forms.map((item, index) => {
              if (index === activeTab) {
                return (
                  <Styled.Form
                    key={item.name}
                    action={item.action}
                    method="post"
                  >
                    <label htmlFor={`${id}-invoice`} hidden>
                      Invoice Number
                    </label>
                    <Styled.Input
                      id={`${id}-invoice`}
                      name="invoiceNumber"
                      autoComplete="off"
                      type="text"
                      maxLength="10"
                      placeholder="Invoice No."
                      required
                    />
                    <label htmlFor={id} hidden>
                      {item.placeholder.replace('No.', 'Number')}
                    </label>
                    <Styled.Input
                      id={id}
                      name={item.name}
                      autoComplete="off"
                      type="text"
                      maxLength="10"
                      placeholder={item.placeholder}
                      required
                    />
                    <Styled.SubmitButton
                      type="submit"
                      title="Submit"
                      aria-label="Submit"
                    >
                      <MaterialIcon iconSize="32px" iconColor="white">
                        chevron_right
                      </MaterialIcon>
                    </Styled.SubmitButton>
                  </Styled.Form>
                );
              }
              return undefined;
            })}
          </motion.div>
        )}
      </AnimatePresence>
    </Styled.Container>
  );
}
