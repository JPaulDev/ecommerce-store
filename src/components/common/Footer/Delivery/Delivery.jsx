import { useState, useEffect } from 'react';
import { motion, AnimatePresence } from 'framer-motion';
import { useMediaQuery } from '../../../../hooks';
import { LabelledInput } from '../../../ui';
import { ChevronRight } from '../../../icons';
import { Heading, DropdownButton } from '../styles';
import * as Styled from './styles';

const forms = [
  {
    action: 'https://pc-connect.co.uk/tracking/trackorder',
    placeholder: 'Postcode',
    name: 'postcode',
  },
  {
    action: 'https://pc-connect.co.uk/tracking/trackreturn',
    placeholder: 'RMA No.',
    name: 'rmaNumber',
  },
  {
    action: 'https://pc-connect.co.uk/tracking/trackbuild',
    placeholder: 'Build No.',
    name: 'buildNumber',
  },
];

const variants = {
  open: {
    overflow: 'hidden',
    height: 'auto',
    transition: {
      ease: 'easeInOut',
      duration: 0.5,
    },
  },
  closed: {
    overflow: 'hidden',
    height: 0,
    transition: {
      ease: 'easeInOut',
      duration: 0.5,
    },
  },
};

export default function Delivery() {
  const [activeTab, setActiveTab] = useState(0);
  const [isOpen, setIsOpen] = useState(false);
  const isMatch = useMediaQuery('(min-width: 650px)');
  const heading = 'Delivery / Tracking Information';

  const handleSetActiveTab = (index) => setActiveTab(index);

  const handleOpen = () => setIsOpen(!isOpen);

  // Prevents the menu being stuck closed when resizing from a smaller window
  // width.
  useEffect(() => {
    isMatch ? setIsOpen(true) : setIsOpen(false);
  }, [isMatch]);

  return (
    <Styled.Container>
      {isMatch ? (
        <Heading>{heading}</Heading>
      ) : (
        <DropdownButton
          as="button"
          type="button"
          isOpen={isOpen}
          aria-expanded={isOpen}
          aria-controls="delivery"
          onClick={handleOpen}
        >
          {heading}
          <ChevronRight width={15} height={20} />
        </DropdownButton>
      )}
      <AnimatePresence>
        {isOpen && (
          <motion.div
            id="delivery"
            variants={variants}
            initial="closed"
            animate="open"
            exit="closed"
          >
            <Styled.Text>Order now to receive by tomorrow.</Styled.Text>
            <Styled.ButtonContainer>
              {['Order', 'Return', 'System'].map((text, index) => (
                <Styled.TabButton
                  key={text}
                  active={index === activeTab}
                  onClick={() => handleSetActiveTab(index)}
                  type="button"
                >
                  {text}
                </Styled.TabButton>
              ))}
            </Styled.ButtonContainer>
            {forms.map((item, index) => {
              if (index === activeTab) {
                return (
                  <Styled.Form
                    key={item.name}
                    action={item.action}
                    method="post"
                  >
                    <LabelledInput
                      label="Invoice Number:"
                      styles={Styled.inputStyles}
                      name="invoiceNumber"
                      type="text"
                      placeholder="Invoice No."
                      autoComplete="off"
                      maxLength="10"
                      aria-required
                      required
                    />
                    <LabelledInput
                      label={`${item.placeholder.replace('No.', 'Number')}:`}
                      styles={Styled.inputStyles}
                      name={item.name}
                      type="text"
                      placeholder={item.placeholder}
                      autoComplete="off"
                      maxLength="10"
                      aria-required
                      required
                    />
                    <Styled.SubmitButton
                      type="submit"
                      title="Submit"
                      aria-label="Submit"
                    >
                      <ChevronRight width={14} height={19} fill="white" />
                    </Styled.SubmitButton>
                  </Styled.Form>
                );
              }
              return undefined;
            })}
          </motion.div>
        )}
      </AnimatePresence>
    </Styled.Container>
  );
}
