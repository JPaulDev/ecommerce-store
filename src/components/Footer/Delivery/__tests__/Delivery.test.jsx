import { render, screen, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import Delivery from '../index';

describe('Delivery component', () => {
  it('renders an order tab where order tracking can be checked', () => {
    const tree = renderer.create(<Delivery />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a return tab where returns status can be tracked', () => {
    const component = renderer.create(<Delivery />);

    act(() => {
      component.root.findAllByType('button')[1].props.onClick();
    });

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders a system build tab where build progress can be tracked', () => {
    const component = renderer.create(<Delivery />);

    act(() => {
      component.root.findAllByType('button')[2].props.onClick();
    });

    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe('Delivery component text inputs', () => {
  it('correctly records user input', () => {
    render(<Delivery />);
    const invoice = screen.getByLabelText(/invoice number/i);
    const postcode = screen.getByLabelText(/postcode/i);

    userEvent.type(invoice, '12345');
    userEvent.type(postcode, 'ABCDE');

    expect(invoice).toHaveValue('12345');
    expect(postcode).toHaveValue('ABCDE');
  });
});
