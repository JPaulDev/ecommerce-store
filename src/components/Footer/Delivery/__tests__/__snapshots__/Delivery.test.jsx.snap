// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Delivery component renders a return tab where returns status can be tracked 1`] = `
<div
  className="sc-gKXOVf NsfSX"
>
  <h3
    className="sc-eCYdqJ eSKKko"
  >
    Delivery / Tracking Information
  </h3>
  <div
    id="delivery"
    style={
      Object {
        "height": "auto",
      }
    }
  >
    <p
      className="sc-iBkjds jgsQoG"
    >
      Order now to receive by tomorrow.
    </p>
    <div
      className="sc-ftvSup ehGKsD"
    >
      <button
        className="sc-papXJ hWBmUK"
        onClick={[Function]}
        type="button"
      >
        Order
      </button>
      <button
        className="sc-papXJ fdnhZi"
        onClick={[Function]}
        type="button"
      >
        Return
      </button>
      <button
        className="sc-papXJ hWBmUK"
        onClick={[Function]}
        type="button"
      >
        System
      </button>
    </div>
    <form
      action="https://pc-connect.co.uk/tracking/trackreturn"
      className="sc-jqUVSM laBddG"
      method="post"
    >
      <label
        hidden={true}
        htmlFor=":r3:-invoice"
      >
        Invoice Number
      </label>
      <input
        autoComplete="off"
        className="sc-kDDrLX bFbYeH"
        id=":r3:-invoice"
        maxLength="10"
        name="invoiceNumber"
        placeholder="Invoice No."
        required={true}
        type="text"
      />
      <label
        hidden={true}
        htmlFor=":r3:"
      >
        RMA Number
      </label>
      <input
        autoComplete="off"
        className="sc-kDDrLX bFbYeH"
        id=":r3:"
        maxLength="10"
        name="rmaNumber"
        placeholder="RMA No."
        required={true}
        type="text"
      />
      <button
        aria-label="Submit"
        className="sc-iqcoie cNpgvc"
        title="Submit"
        type="submit"
      >
        <i
          className="sc-bczRLJ kVcBsA material-icons"
        >
          chevron_right
        </i>
      </button>
    </form>
  </div>
</div>
`;

exports[`Delivery component renders a system build tab where build progress can be tracked 1`] = `
<div
  className="sc-gKXOVf NsfSX"
>
  <h3
    className="sc-eCYdqJ eSKKko"
  >
    Delivery / Tracking Information
  </h3>
  <div
    id="delivery"
    style={
      Object {
        "height": "auto",
      }
    }
  >
    <p
      className="sc-iBkjds jgsQoG"
    >
      Order now to receive by tomorrow.
    </p>
    <div
      className="sc-ftvSup ehGKsD"
    >
      <button
        className="sc-papXJ hWBmUK"
        onClick={[Function]}
        type="button"
      >
        Order
      </button>
      <button
        className="sc-papXJ hWBmUK"
        onClick={[Function]}
        type="button"
      >
        Return
      </button>
      <button
        className="sc-papXJ fdnhZi"
        onClick={[Function]}
        type="button"
      >
        System
      </button>
    </div>
    <form
      action="https://pc-connect.co.uk/tracking/trackbuild"
      className="sc-jqUVSM laBddG"
      method="post"
    >
      <label
        hidden={true}
        htmlFor=":r6:-invoice"
      >
        Invoice Number
      </label>
      <input
        autoComplete="off"
        className="sc-kDDrLX bFbYeH"
        id=":r6:-invoice"
        maxLength="10"
        name="invoiceNumber"
        placeholder="Invoice No."
        required={true}
        type="text"
      />
      <label
        hidden={true}
        htmlFor=":r6:"
      >
        Build Number
      </label>
      <input
        autoComplete="off"
        className="sc-kDDrLX bFbYeH"
        id=":r6:"
        maxLength="10"
        name="buildNumber"
        placeholder="Build No."
        required={true}
        type="text"
      />
      <button
        aria-label="Submit"
        className="sc-iqcoie cNpgvc"
        title="Submit"
        type="submit"
      >
        <i
          className="sc-bczRLJ kVcBsA material-icons"
        >
          chevron_right
        </i>
      </button>
    </form>
  </div>
</div>
`;

exports[`Delivery component renders an order tab where order tracking can be checked 1`] = `
<div
  className="sc-gKXOVf NsfSX"
>
  <h3
    className="sc-eCYdqJ eSKKko"
  >
    Delivery / Tracking Information
  </h3>
  <div
    id="delivery"
    style={
      Object {
        "height": "auto",
      }
    }
  >
    <p
      className="sc-iBkjds jgsQoG"
    >
      Order now to receive by tomorrow.
    </p>
    <div
      className="sc-ftvSup ehGKsD"
    >
      <button
        className="sc-papXJ fdnhZi"
        onClick={[Function]}
        type="button"
      >
        Order
      </button>
      <button
        className="sc-papXJ hWBmUK"
        onClick={[Function]}
        type="button"
      >
        Return
      </button>
      <button
        className="sc-papXJ hWBmUK"
        onClick={[Function]}
        type="button"
      >
        System
      </button>
    </div>
    <form
      action="https://pc-connect.co.uk/tracking/trackorder"
      className="sc-jqUVSM laBddG"
      method="post"
    >
      <label
        hidden={true}
        htmlFor=":r0:-invoice"
      >
        Invoice Number
      </label>
      <input
        autoComplete="off"
        className="sc-kDDrLX bFbYeH"
        id=":r0:-invoice"
        maxLength="10"
        name="invoiceNumber"
        placeholder="Invoice No."
        required={true}
        type="text"
      />
      <label
        hidden={true}
        htmlFor=":r0:"
      >
        Postcode
      </label>
      <input
        autoComplete="off"
        className="sc-kDDrLX bFbYeH"
        id=":r0:"
        maxLength="10"
        name="postcode"
        placeholder="Postcode"
        required={true}
        type="text"
      />
      <button
        aria-label="Submit"
        className="sc-iqcoie cNpgvc"
        title="Submit"
        type="submit"
      >
        <i
          className="sc-bczRLJ kVcBsA material-icons"
        >
          chevron_right
        </i>
      </button>
    </form>
  </div>
</div>
`;
